module/x/peggy/keeper/attestation.go

```diff
@@ -18,12 +18,56 @@ import (
 //  - `observed` attestations are processed for state transition
 //  - the process result is stored with the attestion
 //  - an `observation` event is emitted
+//
+// Jehan's note: AddClaim is called every time an eth signer sends in a claim. Each time:
+// - it adds the eth signer's vote to the claim using "tryAttestation". The votes are tallied in a struct called an
+//   "attestation", and when they pass a threshold, an enum called "Certainty" is updated.
+// - it checks "Certainty", and once the threshold is passed, "processAttestation" is called,
+//   which checks the nonce, updates the nonce, and actually moves tokens.
+//
+// A problem exists, involving two perfectly legitimate, non duplicated bridge deposits. The first could have e.g. nonce 101, and the second nonce 102.
+// If eth signers are not sending claims in to the Cosmos chain in strict nonce order, the second could happen to pass the vote threshold first, resulting
+// in it being impossible to submit the first. This failure is non-obvious, and could happen if one validator's packet containing
+// the claim for the first tx was dropped, or maybe just because of quirks of the Cosmos gossip network.
+//
+// To fix this here is a possible solution:
+// - Enforce strict nonce ordering on claims from a given validator. Do not let a deposit be submitted unless the nonce is exactly one
+//   higher than the last claim of the last deposit. This could be added with a nonce check in storeClaim.
+//   Eth signers will need to know how to avoid this check failing, either by knowing how to retry earlier transactions,
+//   or by not sending a claim before the last claim was successfully submitted.
+// - Don't cause a state transition (e.g. sending tokens) immediately once a deposit has reached the threshold of votes (is "observed"). Every block,
+//   attempt to play the "observed" deposit claims back and perform the state transitions. Stop if the nonce ever increases by more than 1.
+//   The unused deposit claims stick around and can be retried next block. Eth signers still need to know how to retry claims, otherwise they will be slashed.
+//
+// Jehan's note: The nonce logic for the batches is completely different. As shown above, we have to reject deposit claims
+// if all earlier deposit claims have not yet entered the system.
+// Batch claims are totally different. For batches, we need to free the transactions of earlier batches when a later batch claims
+// comes in. So given that this function is common to both types of claims, it is a bad place to implement any batch logic.
+// We will have to think about how to split the code. We may want to make two versions of this whole function.
+//
+// Let's say there are two batches in the batch pool. Batches with nonce 101 and 102. Both are successfully submitted to the
```

Comment:
this can be prevented by having the eth_signer enforce ordering. This is already a trusted role so I don't see too much of a difference adding this responsibility.

Reply:
We covered this in chat earlier, but that seems very brittle and non obvious